README.md
# go-atari-cli

[![GoDoc](https://godoc.org/github.com/thewizardplusplus/go-atari-cli?status.svg)](https://godoc.org/github.com/thewizardplusplus/go-atari-cli)
[![Go Report Card](https://goreportcard.com/badge/github.com/thewizardplusplus/go-atari-cli)](https://goreportcard.com/report/github.com/thewizardplusplus/go-atari-cli)
[![Build Status](https://travis-ci.org/thewizardplusplus/go-atari-cli.svg?branch=master)](https://travis-ci.org/thewizardplusplus/go-atari-cli)
[![codecov](https://codecov.io/gh/thewizardplusplus/go-atari-cli/branch/master/graph/badge.svg)](https://codecov.io/gh/thewizardplusplus/go-atari-cli)

The [Atari Go](https://senseis.xmp.net/?AtariGo) program with a terminal-based interface.

_**Disclaimer:** this program was written directly on an Android smartphone with the AnGoIde IDE._

## Features

- displaying a board:
  - by symbols (to choose):
    - ASCII;
    - Unicode;
  - by colors (to choose):
    - monochrome;
    - colorful;
  - by size (to choose):
    - terse;
    - wide;
  - by decor (to choose):
    - without the board grid;
    - with the board grid;
  - misc.:
    - marking searching process;
- interacting via text commands (moves in [Smart Game Format](https://senseis.xmp.net/?SGF));
- options:
  - initial position in [Smart Game Format](https://senseis.xmp.net/?SGF);
  - human color (i.e. a computer can move first):
    - support automatic random selecting (optional);
  - move searching restrictions:
    - passes of tree building;
    - duration of tree building;
  - optimization via parallel move searching:
    - parallel game simulating:
      - of a single node child;
      - of all node children;
    - parallel tree building;
  - displaying:
    - switching between ASCII/Unicode modes;
    - switching between monochrome/colorful modes;
    - switching between terse/wide modes;
    - switching between modes without/with the board grid.

## Installation

```
$ go get github.com/thewizardplusplus/go-atari-cli/...
```

## Usage

```
$ go-atari-cli -h | -help | --help
$ go-atari-cli [options]
```

Options:

- `-h`, `-help`, `--help` &mdash; show the help message and exit;
- `-sgf STRING` &mdash; board in [Smart Game Format](https://senseis.xmp.net/?SGF) (default: empty board 5x5);
- `-humanColor {random|black|white}` &mdash; human color (default: `random`);
- `-passes INTEGER` &mdash; building passes (default: `1000`);
- `-duration DURATION` &mdash; building duration (e.g. `72h3m0.5s`; default: `10s`);
- `-parallelSimulator` &mdash; use parallel game simulating of a single node child (default: `false`; for inverting use `-parallelSimulator` or `-parallelSimulator=true`);
- `-parallelBulkySimulator` &mdash; use parallel game simulating of all node children (default: `false`; for inverting use `-parallelBulkySimulator` or `-parallelBulkySimulator=true`);
- `-parallelBuilder` &mdash; use parallel tree building (default: `true`; for inverting use `-parallelBuilder=false`);
- `-unicode` &mdash; use Unicode to display stones (default: `true`; for inverting use `-unicode=false`);
- `-colorful` &mdash; use colors to display stones (default: `true`; for inverting use `-colorful=false`);
- `-blackColor INTEGER` &mdash; SGR parameter for ANSI escape sequences for setting a color of black stones (default: `34`; see for details: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit);
- `-whiteColor INTEGER` &mdash; SGR parameter for ANSI escape sequences for setting a color of white stones (default: `31`; see for details: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit);
- `-wide` &mdash; display the board wide (default: `true`; for inverting use `-wide=false`);
- `-grid` &mdash; display the board grid (default: `true`; for inverting use `-grid=false`).

## Examples

`ascii.DecodeColor()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-atari-cli/encoding/ascii"
)

func main() {
	color, _ := ascii.DecodeColor("white")
	fmt.Printf("%v\n", color)

	// Output: 1
}
```

`ascii.EncodeColor()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-atari-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-atari-models"
)

func main() {
	color := ascii.EncodeColor(models.White)
	fmt.Printf("%v\n", color)

	// Output: white
}
```

`ascii.StoneStorageEncoder.EncodeStoneStorage()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-atari-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-atari-models"
	"github.com/thewizardplusplus/go-atari-models/encoding/sgf"
)

func main() {
	stoneEncoder := func(color models.Color) string {
		return string(sgf.EncodeColor(color))
	}
	placeholders := ascii.Placeholders{
		Crosshairs: "+",
	}

	// +-+-+-+-+-+
	// | | | | | |
	// +-+-+-+-+-+
	// | |B|W| | |
	// +-+-+-+-+-+
	// |B|W| |W| |
	// +-+-+-+-+-+
	// | |B|W| | |
	// +-+-+-+-+-+
	// | | | | | |
	// +-+-+-+-+-+
	board := models.NewBoard(models.Size{Width: 5, Height: 5})
	for _, move := range []models.Move{
		{Color: models.Black, Point: models.Point{Column: 1, Row: 1}},
		{Color: models.White, Point: models.Point{Column: 2, Row: 1}},
		{Color: models.Black, Point: models.Point{Column: 0, Row: 2}},
		{Color: models.White, Point: models.Point{Column: 1, Row: 2}},
		{Color: models.White, Point: models.Point{Column: 3, Row: 2}},
		{Color: models.Black, Point: models.Point{Column: 1, Row: 3}},
		{Color: models.White, Point: models.Point{Column: 2, Row: 3}},
	} {
		board = board.ApplyMove(move)
	}

	boardEncoder :=
		ascii.NewStoneStorageEncoder(stoneEncoder, placeholders, ascii.Margins{}, 1)
	fmt.Printf("%v\n", boardEncoder.EncodeStoneStorage(board))

	// Output:
	// e+++++
	// d+BW++
	// cBW+W+
	// b+BW++
	// a+++++
	//  abcde
}
```

`ascii.StoneStorageEncoder.EncodeStoneStorage()` with margins:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-atari-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-atari-models"
	"github.com/thewizardplusplus/go-atari-models/encoding/sgf"
)

func main() {
	stoneEncoder := func(color models.Color) string {
		return string(sgf.EncodeColor(color))
	}
	placeholders := ascii.Placeholders{
		HorizontalLine: " ",
		Crosshairs:     "+",
	}
	margins := ascii.Margins{
		Stone: ascii.StoneMargins{
			HorizontalMargins: ascii.HorizontalMargins{
				Left: 1,
			},
		},
		Legend: ascii.LegendMargins{
			Row: ascii.HorizontalMargins{
				Right: 1,
			},
		},
	}

	// +-+-+-+-+-+
	// | | | | | |
	// +-+-+-+-+-+
	// | |B|W| | |
	// +-+-+-+-+-+
	// |B|W| |W| |
	// +-+-+-+-+-+
	// | |B|W| | |
	// +-+-+-+-+-+
	// | | | | | |
	// +-+-+-+-+-+
	board := models.NewBoard(models.Size{Width: 5, Height: 5})
	for _, move := range []models.Move{
		{Color: models.Black, Point: models.Point{Column: 1, Row: 1}},
		{Color: models.White, Point: models.Point{Column: 2, Row: 1}},
		{Color: models.Black, Point: models.Point{Column: 0, Row: 2}},
		{Color: models.White, Point: models.Point{Column: 1, Row: 2}},
		{Color: models.White, Point: models.Point{Column: 3, Row: 2}},
		{Color: models.Black, Point: models.Point{Column: 1, Row: 3}},
		{Color: models.White, Point: models.Point{Column: 2, Row: 3}},
	} {
		board = board.ApplyMove(move)
	}

	boardEncoder :=
		ascii.NewStoneStorageEncoder(stoneEncoder, placeholders, margins, 1)
	fmt.Printf("%v\n", boardEncoder.EncodeStoneStorage(board))

	// Output:
	// e  + + + + +
	// d  + B W + +
	// c  B W + W +
	// b  + B W + +
	// a  + + + + +
	//    a b c d e
}
```

`ascii.StoneStorageEncoder.EncodeStoneStorage()` with a grid:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-atari-cli/encoding/ascii"
	models "github.com/thewizardplusplus/go-atari-models"
	"github.com/thewizardplusplus/go-atari-models/encoding/sgf"
)

func main() {
	stoneEncoder := func(color models.Color) string {
		return string(sgf.EncodeColor(color))
	}
	placeholders := ascii.Placeholders{
		HorizontalLine: "-",
		VerticalLine:   "|",
		Crosshairs:     "+",
	}
	margins := ascii.Margins{
		Stone: ascii.StoneMargins{
			HorizontalMargins: ascii.HorizontalMargins{
				Left: 1,
			},
			VerticalMargins: ascii.VerticalMargins{
				Bottom: 1,
			},
		},
		Legend: ascii.LegendMargins{
			Row: ascii.HorizontalMargins{
				Right: 1,
			},
		},
	}

	// +-+-+-+-+-+
	// | | | | | |
	// +-+-+-+-+-+
	// | |B|W| | |
	// +-+-+-+-+-+
	// |B|W| |W| |
	// +-+-+-+-+-+
	// | |B|W| | |
	// +-+-+-+-+-+
	// | | | | | |
	// +-+-+-+-+-+
	board := models.NewBoard(models.Size{Width: 5, Height: 5})
	for _, move := range []models.Move{
		{Color: models.Black, Point: models.Point{Column: 1, Row: 1}},
		{Color: models.White, Point: models.Point{Column: 2, Row: 1}},
		{Color: models.Black, Point: models.Point{Column: 0, Row: 2}},
		{Color: models.White, Point: models.Point{Column: 1, Row: 2}},
		{Color: models.White, Point: models.Point{Column: 3, Row: 2}},
		{Color: models.Black, Point: models.Point{Column: 1, Row: 3}},
		{Color: models.White, Point: models.Point{Column: 2, Row: 3}},
	} {
		board = board.ApplyMove(move)
	}

	boardEncoder :=
		ascii.NewStoneStorageEncoder(stoneEncoder, placeholders, margins, 1)
	fmt.Printf("%v\n", boardEncoder.EncodeStoneStorage(board))

	// Output:
	// e -+-+-+-+-+
	//    | | | | |
	// d -+-B-W-+-+
	//    | | | | |
	// c -B-W-+-W-+
	//    | | | | |
	// b -+-B-W-+-+
	//    | | | | |
	// a -+-+-+-+-+
	//    | | | | |
	//    a b c d e
}
```

`unicode.EncodeStone()`:

```go
package main

import (
	"fmt"

	"github.com/thewizardplusplus/go-atari-cli/encoding/unicode"
	models "github.com/thewizardplusplus/go-atari-models"
)

func main() {
	stone := unicode.EncodeStone(models.White)
	fmt.Printf("%v\n", stone)

	// Output: ○
}
```

## License

The MIT License (MIT)

Copyright &copy; 2020 thewizardplusplus


CHANGE_LOG.md
# Change Log

## [v1.4](https://github.com/thewizardplusplus/go-atari-cli/tree/v1.4) (2020-07-04)

- fixing:
  - code style;
  - command-line flags;
- adding usage examples;
- improving repository decor.

## [v1.3](https://github.com/thewizardplusplus/go-atari-cli/tree/v1.3) (2020-05-30)

- using:
  - board interface;
  - separate move generator.

## [v1.2](https://github.com/thewizardplusplus/go-atari-cli/tree/v1.2) (2020-05-18)

- displaying a board:
  - changing ASCII symbols for stones displaying;
  - by decor (to choose):
    - without the board grid;
    - with the board grid.

## [v1.1](https://github.com/thewizardplusplus/go-atari-cli/tree/v1.1) (2020-05-12)

- displaying a board:
  - swapping Unicode symbols for stones displaying;
- options:
  - specifying an initial position in [Smart Game Format](https://senseis.xmp.net/?SGF).

## [v1.0](https://github.com/thewizardplusplus/go-atari-cli/tree/v1.0) (2020-05-06)


cmd/go-atari-cli/main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"log"
	"math"
	"math/rand"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/thewizardplusplus/go-atari-cli/encoding/ascii"
	"github.com/thewizardplusplus/go-atari-cli/encoding/unicode"
	climodels "github.com/thewizardplusplus/go-atari-cli/models"
	models "github.com/thewizardplusplus/go-atari-models"
	"github.com/thewizardplusplus/go-atari-models/encoding/sgf"
	"github.com/thewizardplusplus/go-atari-montecarlo/builders"
	"github.com/thewizardplusplus/go-atari-montecarlo/builders/terminators"
	"github.com/thewizardplusplus/go-atari-montecarlo/searchers"
	"github.com/thewizardplusplus/go-atari-montecarlo/selectors"
	"github.com/thewizardplusplus/go-atari-montecarlo/selectors/scorers"
	"github.com/thewizardplusplus/go-atari-montecarlo/simulators"
	"github.com/thewizardplusplus/go-atari-montecarlo/simulators/bulky"
	"github.com/thewizardplusplus/go-atari-montecarlo/tree"
)

const (
	ucbFactor = math.Sqrt2
)

// nolint: gochecknoglobals
var (
	asciiPlaceholders = ascii.Placeholders{
		HorizontalLine: "-",
		VerticalLine:   "|",
		Crosshairs:     "+",
	}
	unicodePlaceholders = ascii.Placeholders{
		HorizontalLine: "\u2500",
		VerticalLine:   "\u2502",
		Crosshairs:     "\u253c",
	}

	baseWideMargins = ascii.Margins{
		Legend: ascii.LegendMargins{
			Column: ascii.VerticalMargins{
				Top: 1,
			},
			Row: ascii.HorizontalMargins{
				Right: 1,
			},
		},
		Board: ascii.VerticalMargins{
			Top:    1,
			Bottom: 1,
		},
	}
	wideStoneMargins = ascii.StoneMargins{
		HorizontalMargins: ascii.HorizontalMargins{
			Left: 1,
		},
		VerticalMargins: ascii.VerticalMargins{
			Bottom: 1,
		},
	}
	extraWideStoneMargins = ascii.StoneMargins{
		HorizontalMargins: ascii.HorizontalMargins{
			Left:  1,
			Right: 1,
		},
		VerticalMargins: ascii.VerticalMargins{
			Top:    1,
			Bottom: 1,
		},
	}
)

type colorCodeGroup map[models.Color]int

func colorize(
	text string,
	color models.Color,
	colorsCodes colorCodeGroup,
) string {
	return setTTYMode(colorsCodes[color]) + text + setTTYMode(0)
}

func setTTYMode(mode int) string {
	return fmt.Sprintf("\x1b[%dm", mode)
}

type searchSettings struct {
	maximalPass            int
	maximalDuration        time.Duration
	parallelSimulator      bool
	parallelBulkySimulator bool
	parallelBuilder        bool
}

func search(
	storage models.StoneStorage,
	color models.Color,
	settings searchSettings,
) (models.Move, error) {
	generator := models.MoveGenerator{}

	randomSelector := selectors.RandomMoveSelector{}
	generalSelector := selectors.MaximalNodeSelector{
		NodeScorer: scorers.UCBScorer{
			Factor: ucbFactor,
		},
	}

	var simulator simulators.Simulator // nolint: staticcheck
	simulator = simulators.RolloutSimulator{
		MoveGenerator: generator,
		MoveSelector:  randomSelector,
	}
	if settings.parallelSimulator {
		simulator = simulators.ParallelSimulator{
			Simulator:   simulator,
			Concurrency: runtime.NumCPU(),
		}
	}

	var bulkySimulator builders.BulkySimulator
	if !settings.parallelBulkySimulator {
		bulkySimulator = bulky.FirstNodeSimulator{
			Simulator: simulator,
		}
	} else {
		bulkySimulator = bulky.AllNodesSimulator{
			Simulator: simulator,
		}
	}

	var builder builders.Builder
	terminator := terminators.NewGroupTerminator(
		terminators.NewPassTerminator(settings.maximalPass),
		terminators.NewTimeTerminator(time.Now, settings.maximalDuration),
	)
	builder = builders.IterativeBuilder{
		Builder: builders.TreeBuilder{
			NodeSelector:  generalSelector,
			MoveGenerator: generator,
			Simulator:     bulkySimulator,
		},
		Terminator: terminator,
	}
	if settings.parallelBuilder {
		builder = builders.ParallelBuilder{
			Builder:     builder,
			Concurrency: runtime.NumCPU(),
		}
	}

	root := &tree.Node{
		Move:    models.NewPreliminaryMove(color),
		Storage: storage,
	}
	searcher := searchers.MoveSearcher{
		MoveGenerator: generator,
		Builder:       builder,
		NodeSelector:  generalSelector,
	}
	node, err := searcher.SearchMove(root)
	if err != nil {
		return models.Move{}, err
	}

	return node.Move, nil
}

func check(storage models.StoneStorage, color models.Color) error {
	generator := models.MoveGenerator{}
	_, err := generator.LegalMoves(storage, models.NewPreliminaryMove(color))
	return err // don't wrap
}

func writePrompt(
	storageEncoder ascii.StoneStorageEncoder,
	storage models.StoneStorage,
	color models.Color,
	side climodels.Side,
) error {
	text := storageEncoder.EncodeStoneStorage(storage)
	fmt.Println(text)

	if err := check(storage, color); err != nil {
		return err // don't wrap
	}

	var mark string
	if side == climodels.Searcher {
		mark = "(searching) "
	}
	prompt := makePrompt(color, mark)
	fmt.Print(prompt) // don't break the line

	return nil
}

func makePrompt(color models.Color, data interface{}) string {
	prompt := ascii.EncodeColor(color)
	return fmt.Sprintf("%s> %v", prompt, data)
}

func readMove(
	reader *bufio.Reader,
	storageEncoder ascii.StoneStorageEncoder,
	storage models.StoneStorage,
	color models.Color,
	side climodels.Side,
) (models.Move, error) {
	if err := writePrompt(storageEncoder, storage, color, side); err != nil {
		return models.Move{}, err // don't wrap
	}

	text, err := reader.ReadString('\n')
	if err != nil && err != io.EOF {
		return models.Move{}, fmt.Errorf("unable to read the move: %s", err)
	}

	text = strings.TrimSuffix(text, "\n")
	point, err := sgf.DecodePoint(text)
	if err != nil {
		return models.Move{}, fmt.Errorf("unable to decode the point: %s", err)
	}

	move := models.Move{
		Color: color,
		Point: point,
	}
	if err := storage.CheckMove(move); err != nil {
		return models.Move{}, fmt.Errorf("incorrect move: %s", err)
	}

	return move, nil
}

func searchMove(
	storageEncoder ascii.StoneStorageEncoder,
	storage models.StoneStorage,
	color models.Color,
	side climodels.Side,
	settings searchSettings,
) (models.Move, error) {
	if err := writePrompt(storageEncoder, storage, color, side); err != nil {
		return models.Move{}, err // don't wrap
	}

	return search(storage, color, settings)
}

func main() {
	rand.Seed(time.Now().UnixNano())

	storageInSGF := flag.String(
		"sgf",
		"",
		"board in SGF (default: empty board 5x5)",
	)
	humanColor := flag.String(
		"humanColor",
		"random",
		"human color (allowed: random, black, white)",
	)
	pass := flag.Int("pass", 1000, "building pass")
	duration := flag.Duration(
		"duration",
		10*time.Second,
		"building duration (e.g. 72h3m0.5s)",
	)
	parallelSimulator := flag.Bool(
		"parallelSimulator",
		false,
		"parallel simulator",
	)
	parallelBulkySimulator := flag.Bool(
		"parallelBulkySimulator",
		false,
		"parallel bulky simulator",
	)
	parallelBuilder := flag.Bool("parallelBuilder", true, "parallel builder")
	useUnicode := flag.Bool("unicode", false, "use Unicode to display stones")
	colorful := flag.Bool("colorful", false, "use colors to display stones")
	blackColor := flag.Int(
		"blackColor",
		34, // blue
		"SGR parameter for ANSI escape sequences for setting a color of black stones",
	)
	whiteColor := flag.Int(
		"whiteColor",
		31, // red
		"SGR parameter for ANSI escape sequences for setting a color of white stones",
	)
	wide := flag.Bool("wide", false, "display the board wide")
	grid := flag.Bool("grid", true, "display the board grid")
	flag.Parse()

	storage, err := sgf.DecodeStoneStorage(*storageInSGF, models.NewBoard)
	if err != nil {
		log.Fatal("unable to decode the board: ", err)
	}

	parsedHumanColor, err := ascii.DecodeColor(*humanColor)
	switch {
	case err == nil:
	case *humanColor == "random":
		if rand.Intn(2) == 0 {
			parsedHumanColor = models.Black
		} else {
			parsedHumanColor = models.White
		}
	default:
		log.Fatal("unable to decode the color: ", err)
	}

	var stoneEncoder ascii.StoneEncoder
	var placeholders ascii.Placeholders
	if *useUnicode {
		stoneEncoder = unicode.EncodeStone
		placeholders = unicodePlaceholders
	} else {
		stoneEncoder = func(color models.Color) string {
			return string(sgf.EncodeColor(color))
		}
		placeholders = asciiPlaceholders
	}
	if *colorful {
		baseStoneEncoder := stoneEncoder
		stoneEncoder = func(color models.Color) string {
			text := baseStoneEncoder(color)
			return colorize(text, color, colorCodeGroup{
				models.Black: *blackColor,
				models.White: *whiteColor,
			})
		}
	}
	if !*grid {
		placeholders.HorizontalLine = " "
		placeholders.VerticalLine = " "
	}

	var margins ascii.Margins
	if *wide {
		margins = baseWideMargins

		if *grid {
			margins.Stone = extraWideStoneMargins
		} else {
			margins.Stone = wideStoneMargins
		}
	}

	side := climodels.NewSide(parsedHumanColor)
	reader := bufio.NewReader(os.Stdin)
	storageEncoder := ascii.NewStoneStorageEncoder(
		stoneEncoder,
		placeholders,
		margins,
		1,
	)
	settings := searchSettings{
		maximalPass:            *pass,
		maximalDuration:        *duration,
		parallelSimulator:      *parallelSimulator,
		parallelBulkySimulator: *parallelBulkySimulator,
		parallelBuilder:        *parallelBuilder,
	}
loop:
	for {
		var currentColor models.Color
		var move models.Move
		var err error
		switch side {
		case climodels.Human:
			currentColor = parsedHumanColor
			move, err = readMove(reader, storageEncoder, storage, currentColor, side)
		case climodels.Searcher:
			currentColor = parsedHumanColor.Negative()
			move, err = searchMove(storageEncoder, storage, currentColor, side, settings)
			if err == nil {
				text := sgf.EncodePoint(move.Point)
				fmt.Println(text)
			}
		}
		switch err {
		case nil:
		case models.ErrAlreadyLoss, models.ErrAlreadyWin:
			prompt := makePrompt(currentColor, err)
			fmt.Println(prompt)

			break loop
		default:
			log.Print("error: ", err)
			continue loop
		}

		storage = storage.ApplyMove(move)
		side = side.Invert()
	}
}
